src/createScope.tsx

```diff
@@ -18,59 +16,62 @@ export const ScopeProvider = ({
+    const atomSet = new Set(atoms);
+    const createScopedAtom = <T extends AnyWritableAtom>(
+      anAtom: T,
+      delegate: boolean,
+    ): T => {
+      const getAtom = <A extends AnyAtom>(
+        thisArg: AnyAtom,
+        orig: AnyAtom,
+        target: A,
+      ): A => {
+        if (target === thisArg) {
+          return delegate ? getParentScopedAtom(orig as A) : target;
+        }
+        return getScopedAtom(target);
+      };
+      const scopedAtom: typeof anAtom = {
+        ...anAtom,
+        ...('read' in anAtom && {
+          read(get, opts) {
+            return anAtom.read.call(
+              this,
+              (a) => get(getAtom(this, anAtom, a)),
+              opts,
+            );
+          },
+        }),
+        ...('write' in anAtom && {
+          write(get, set, ...args) {
+            return anAtom.write.call(
+              this,
+              (a) => get(getAtom(this, anAtom, a)),
+              (a, ...v) => set(getAtom(this, anAtom, a), ...v),
+              ...args,
+            );
+          },
+        }),
+      };
+      return scopedAtom;
+    };
 
-  const createScopedAtom = <T extends AnyWritableAtom>(
-    anAtom: T,
-    delegate: boolean,
-  ): T => {
-    const getAtom = <A extends AnyAtom>(
-      thisArg: AnyAtom,
-      orig: AnyAtom,
-      target: A,
-    ): A => {
-      if (target === thisArg) {
-        return delegate ? getParentScopedAtom(orig as A) : target;
+    const getScopedAtom: GetScopedAtom = (anAtom) => {
+      let scopedAtom = mapping.get(anAtom);
+      if (!scopedAtom) {
+        scopedAtom = atomSet.has(anAtom)
+          ? createScopedAtom(anAtom as unknown as AnyWritableAtom, false)
+          : createScopedAtom(anAtom as unknown as AnyWritableAtom, true);
+        mapping.set(anAtom, scopedAtom);
       }
-      return getScopedAtom(target);
-    };
-    const scopedAtom: typeof anAtom = {
-      ...anAtom,
-      ...('read' in anAtom && {
-        read(get, opts) {
-          return anAtom.read.call(
-            this,
-            (a) => get(getAtom(this, anAtom, a)),
-            opts,
-          );
-        },
-      }),
-      ...('write' in anAtom && {
-        write(get, set, ...args) {
-          return anAtom.write.call(
-            this,
-            (a) => get(getAtom(this, anAtom, a)),
-            (a, ...v) => set(getAtom(this, anAtom, a), ...v),
-            ...args,
-          );
-        },
-      }),
+      return scopedAtom as typeof anAtom;
     };
-    return scopedAtom;
-  };
 
-  const getScopedAtom: GetScopedAtom = (anAtom) => {
-    let scopedAtom = mapping.get(anAtom);
-    if (!scopedAtom) {
-      scopedAtom = atomSet.has(anAtom)
-        ? createScopedAtom(anAtom as unknown as AnyWritableAtom, false)
-        : createScopedAtom(anAtom as unknown as AnyWritableAtom, true);
-      mapping.set(anAtom, scopedAtom);
-    }
-    return scopedAtom as typeof anAtom;
-  };
+    return getScopedAtom;
+  }, [getParentScopedAtom, ...atoms]);
```

Comment:
This dependency array should be treated carefully, whichever of `useMemo` or `memo`. Each independent atom is an dependency, the whole array of atoms is not the dependency.

Reply:
You can't use `...atoms` in deps. The length must not change.

Reply:
What will happen if the set of scoped atoms change 🤔

Reply:
If the length changes, React will complain in DEV mode, and I don't know how production works, maybe ignores the rest, or always changes.

Reply:
I know the dep array's length cannot change. I mean, maybe we should discourage changing the length in ScopeProvider's atoms prop? I'm wondering if it is really needed to change the length.